exercises/Daniel_Murphy/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,151 @@
+#include <iostream>
+#include <fstream>
+#include <map>
+#include <string>
+#include <vector>
+#include <algorithm>
+#include <functional>
+using namespace std;
+
+class ReadFile
+{
+public:
+	std::vector<string> AnalyseLine(string line);
+	std::map<string, int> UpdateMap(std::vector<string> words);
+};
+
+
+std::vector<string> ReadFile::AnalyseLine(string line)
+{
+	std::vector<int> positions = {-1}; // holds positions of all spaces, include the start of the line
+
+	// Remove punctuation that accompanies a space
+	std::vector<char> punctuation = {'?','.',',', '\'', '"', '(', ')', '!', ':', ';'};
+	for (char mark : punctuation)
+	{
+		line.erase(std::remove(line.begin(), line.end(),mark), line.end());
+	}
+
+	// Hyphens don't accompany spaces so replace with a space and split words
+	std::replace( line.begin(), line.end(), '-', ' ');
+
+	// force all letters to case lower
+	std::transform(line.begin(), line.end(), line.begin(), [](unsigned char c){ return std::tolower(c);});
+
+	// Find position of space
+	int pos = static_cast<int>(line.find(" ",0));
+	while(pos != static_cast<int>(string::npos))
+	{
+		positions.push_back(pos);
+		pos = static_cast<int>(line.find(" ",pos+1));
+	}
+	positions.push_back(static_cast<int>(line.length())); // add the end of the line
+
+	// Cut the line by the space positions into words
+	std::vector<string> words;
+	for (auto iter = positions.begin(); iter != positions.end()-1; iter++)
+	{
+		int length = *(iter+1) - ((*iter)+1); // word length = end - start
+		// do not log spaces as words
+		if (length > 4)	{	words.push_back(line.substr(((*iter)+1), length));	}
+	}
+
+	return words;
+}
+
+std::map<string, int> ReadFile::UpdateMap(std::vector<string> words)
+{
+	static std::map<string, int> Counter;
```

Comment:
I agree that the use of `static` seems odd and passing by reference. Just one little thought is that in the new cpp book, when it mentions passing by reference (Page 81) it sums up that "Passing by reference is best avoided." without mentioning any cases when it could be useful.

It obviously doesn't make sense here to pass by const ref, but just out of interest what else would work here? Would using an initialisation constructor to hold the map in an overarching class work better?

Reply:
It's not my place to reply to this, but somehow I ended up subscribed when I posted the other PR

What you could do is encapsulate this into a nice OO object, since we have multiple related methods spread around and we might want to handle edge cases (like updating the count for a single word) in the future:

```
class WordMap{
public:
int getNumberWords(const std::string &word);
void updateWordMap(const std::vector<std::string> &wordList);

private:
using WordMapT = std::map<std::string, int>;
WordMapT counter;
};
```

Then all you need to do is initialise the object type and pass that by ref instead

```
WordMap mappedWords;
mappedWords.updateWordMap(myWords);

```
We can also pass by ref which is very cheap (one pointer length)
```
void printWordMapCounts(const WordMap &allWords){
   const auto &seenWords = allWords.getAllWordsSeen();     // Assume this exists
   for (const auto& word: seenWords){
        std::cout << "I saw " << word << allWords.getNumberWords(word) << " times";
   }
}
```

Reply:
> (Page 81) it sums up that "Passing by reference is best avoided." without mentioning any cases when it could be useful.

What I assume they are trying to say (though I don't have a copy of the book) is that it's a code smell for a few reasons: 

- People will pass in data structures by reference to free functions to modify the data inside. This is programming using C methodology in C++, tie your data and the methods that manipulate it together in a class.
- Modifying parameters forces the reader to look into the implementation to check what happens to their container. If you saw `std::map<std::string, int> getWordCount(const std::string sentence);` you'd know straight away that it's going to make and return a new container (though this gets into the computation cost too)
- People tend to fail the "and / or test" presented in Clean Code, especially having multiple out parameters. "Can you describe your function without using the word *and* or *or*, if not it's likely doing too much and needs chunking down"

Reply:
I'll see if I can get a word in between David's posts :wink:
Funny you noticed that sentence in the new book. I was asked by Martyn\Lamar to review the book with a view to updating the new starter reading list and that line jumped out at me also. I personally think it's a bit strong to say "passing by reference is best avoided". It wouldn't be my first choice of how to pass data into and out of a free function but to say avoid it completely is overshooting a bit in my opinion.
As David's pointed out creating a class is one way to avoid it. I guess if you've more than one function acting on the map it makes the extra lines of code to define the class and create it worthwhile. I believe the exercises are designed to mean that the "new starter" doesn't have to show off their knowledge of classes until exercise 2 so most people don't use a class when answering this question.
So in terms of what you should change to get this PR approved - I don't mind whether you just change the function to take the parameter by reference or create a class. I had commented that your ReadFile class wasn't adding much value but that was because it didn't have a data member. If you follow David's example, it would have a data member.

Reply:
Thanks to both!!!! It's really interesting having comments on my code and the further explanations by David. I'll just change the function to pass by ref for the sake of this exercise and bear in mind that using a class with a data member could be a better option in the future. 
I'm glad to hear that you think the book was a little strong on avoiding pass by reference. It can be hard for someone new to cpp to know how much the advice in the book is a reflection of the community consensus or just the author's opinions. Thankfully they are very clear at many points about avoiding raw pointers, but possibly some other things aren't stated objectively.